File: Kotlin/heapSort.kt
fun heapSort(arr: IntArray) {
    val n = arr.size

    for (i in n / 2 - 1 downTo 0) {
        heapify(arr, n, i)
    }

    for (i in n - 1 downTo 0) {
        val temp = arr[0]
        arr[0] = arr[i]
        arr[i] = temp

        heapify(arr, i, 0)
    }
}

fun heapify(arr: IntArray, n: Int, i: Int) {
    var largest = i
    val left = 2 * i + 1
    val right = 2 * i + 2

    if (left < n && arr[left] > arr[largest]) {
        largest = left
    }

    if (right < n && arr[right] > arr[largest]) {
        largest = right
    }

    if (largest != i) {
        val swap = arr[i]
        arr[i] = arr[largest]
        arr[largest] = swap

        heapify(arr, n, largest)
    }
}

fun printArray(arr: IntArray) {
    val n = arr.size
    for (i in 0 until n) {
        print("${arr[i]} ")
    }
    println()
}

fun main() {
    val arr = intArrayOf(12, 11, 13, 5, 6, 7, 0, -8, 3)

    heapSort(arr)

    println("Sorted array is")
    printArray(arr)
}

File: PHP/heap_sort.php
<?php
function heapify(&$array, $n, $i) {
    $largest = $i;
    $left = 2 * $i + 1;
    $right = 2 * $i + 2;

    if ($left < $n && $array[$left] > $array[$largest]) {
        $largest = $left;
    }

    if ($right < $n && $array[$right] > $array[$largest]) {
        $largest = $right;
    }

    if ($largest != $i) {
        $temp = $array[$i];
        $array[$i] = $array[$largest];
        $array[$largest] = $temp;

        heapify($array, $n, $largest);
    }
}

function heapSort(&$array) {
    $n = count($array);

    for ($i = $n / 2 - 1; $i >= 0; $i--) {
        heapify($array, $n, $i);
    }

    for ($i = $n - 1; $i > 0; $i--) {
        $temp = $array[0];
        $array[0] = $array[$i];
        $array[$i] = $temp;

        heapify($array, $i, 0);
    }
}

$array = [12, 11, 13, 5, 6, 7, 0, -4,-1];
heapSort($array);

echo "Sorted array: ";
echo "[" . implode(", ", $array) . "]";
?>
